src/source_info.rs

```diff
@@ -304,22 +396,83 @@ impl<'i> SourceInfoBuilder<'i> {
         result
     }
 
+    fn get_definition_from_frame(
+        &self,
+        id: &str,
+        frame_id: FrameId,
+        location: &Location,
+    ) -> Option<&Definition> {
+        let mut current_frame_id = Some(frame_id);
+        while let Some(fid) = current_frame_id {
+            // Search definitions in the current scope
+            for definition in self
+                .definitions
+                .iter()
+                .rev()
+                .chain(self.imported_definitions.iter())
+            {
+                let ordering = cmp_range_to_range(&definition.location.range, location.range);
+                if definition.frame_id == fid
+                    && definition.id.as_str() == id
+                    && (ordering == Ordering::Equal || ordering == Ordering::Less)
+                {
+                    return Some(definition);
+                }
+            }
+            // Move to parent scope if available
+            current_frame_id = self.frames.get(fid.0).and_then(|frame| frame.parent);
+        }
+        None
+    }
+
+    fn resolve_references(&mut self) {
+        for unresolved in &self.unresolved_references {
```

Comment:
If you want to be thorough here this could be `in self.unresolved_references.drain(..)`, after the loop the references are resolved so it doesn't make sense to keep them around in `unresolved_references`.